File: backend/src/strategy/trendFollowing.ts
/** EMA(9/21) crossover trend-following entry. */
import { Decimal } from "decimal.js";
import type { OhlcBar } from "../regime/indicators.js";
import type { Signal, Strategy } from "./types.js";

const FAST = 9;
const SLOW = 21;
const SWING_LOOKBACK = 10;

function ema(values: number[], span: number): number[] {
  const alpha = 2 / (span + 1);
  const out: number[] = [values[0]!];
  for (let i = 1; i < values.length; i++) {
    out.push(alpha * values[i]! + (1 - alpha) * out[i - 1]!);
  }
  return out;
}

export class TrendFollowingStrategy implements Strategy {
  strategyId = "trend_following_ema_9_21";

  generateSignal(symbol: string, bars: OhlcBar[]): Signal | null {
    if (bars.length < SLOW + 2) return null;

    const closes = bars.map((b) => b.close);
    const fastEma = ema(closes, FAST);
    const slowEma = ema(closes, SLOW);

    const prevDiff = fastEma[fastEma.length - 2]! - slowEma[slowEma.length - 2]!;
    const currDiff = fastEma[fastEma.length - 1]! - slowEma[slowEma.length - 1]!;

    const swingWindow = bars.slice(-(SWING_LOOKBACK + 1), -1);

    if (prevDiff <= 0 && currDiff > 0) {
      return {
        strategyId: this.strategyId,
        symbol,
        side: "long",
        structureSwingPrice: new Decimal(Math.min(...swingWindow.map((b) => b.low))),
        reason: `${FAST}-EMA crossed above the ${SLOW}-EMA, signaling a new up-trend`,
      };
    }
    if (prevDiff >= 0 && currDiff < 0) {
      return {
        strategyId: this.strategyId,
        symbol,
        side: "short",
        structureSwingPrice: new Decimal(Math.max(...swingWindow.map((b) => b.high))),
        reason: `${FAST}-EMA crossed below the ${SLOW}-EMA, signaling a new down-trend`,
      };
    }
    return null;
  }
}

File: backend/app/strategy/trend_following.py
"""EMA(9/21) crossover trend-following entry."""
from __future__ import annotations

from decimal import Decimal

import pandas as pd

from app.strategy.base import Signal, Strategy

FAST = 9
SLOW = 21
SWING_LOOKBACK = 10


class TrendFollowingStrategy(Strategy):
    strategy_id = "trend_following_ema_9_21"

    def generate_signal(self, symbol: str, df: pd.DataFrame) -> Signal | None:
        if len(df) < SLOW + 2:
            return None

        closes = df["close"]
        fast_ema = closes.ewm(span=FAST, adjust=False).mean()
        slow_ema = closes.ewm(span=SLOW, adjust=False).mean()

        prev_diff = fast_ema.iloc[-2] - slow_ema.iloc[-2]
        curr_diff = fast_ema.iloc[-1] - slow_ema.iloc[-1]

        swing_window = df.iloc[-(SWING_LOOKBACK + 1) : -1]

        if prev_diff <= 0 and curr_diff > 0:
            return Signal(
                strategy_id=self.strategy_id,
                symbol=symbol,
                side="long",
                structure_swing_price=Decimal(str(float(swing_window["low"].min()))),
                reason=f"{FAST}-EMA crossed above the {SLOW}-EMA, signaling a new up-trend",
            )
        if prev_diff >= 0 and curr_diff < 0:
            return Signal(
                strategy_id=self.strategy_id,
                symbol=symbol,
                side="short",
                structure_swing_price=Decimal(str(float(swing_window["high"].max()))),
                reason=f"{FAST}-EMA crossed below the {SLOW}-EMA, signaling a new down-trend",
            )
        return None
